c7n/resources/sns.py

```diff
@@ -252,3 +255,69 @@ def process(self, resources):
                 AttributeValue=json.dumps(policy)
             )
         return results
+
+
+@SNS.filter_registry.register('kms-key')
+class KmsFilter(KmsRelatedFilter):
+
+    RelatedIdsExpression = 'KmsMasterKeyId'
+
+
+@SNS.action_registry.register('set-encryption')
+class SetEncryption(BaseAction):
+    """
+    Set Encryption on SNS Topics
+
+    By default if no key is specified, alias/aws/sns is used
+
+    key can either be a KMS key ARN or an alias, when using aliases
+    prefix the key with 'alias/', i.e.: 'alias/custom/key'.
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: set-sns-topic-encryption
+              resource: sns
+              actions:
+                - type: set-encryption
+                  key: alias/cmk/key
+                  enabled: True
+
+    """
+    schema = type_schema(
+        'set-encryption',
+        enabled={'type': 'boolean'},
+        key={'type': 'string'}
+    )
+
+    permissions = ('sns:SetTopicAttributes', 'kms:DescribeKey',)
+
+    def process(self, resources):
+        sns = local_session(self.manager.session_factory).client('sns')
+
+        if self.data.get('enabled', True):
+            key = self.data.get('key', 'alias/aws/sns')
+            if key != 'alias/aws/sns':
+                kms = local_session(self.manager.session_factory).client('kms')
+                try:
+                    key = kms.describe_key(KeyId=key)
+                except kms.exceptions.NotFoundException:
+                    raise PolicyExecutionError(
+                        "Error when running policy, key: %s not found" % key)
+                key = key.get('KeyMetadata').get('Arn')
+        else:
+            key = ''
+
+        for r in resources:
+            try:
+                sns.set_topic_attributes(
+                    TopicArn=r['TopicArn'],
+                    AttributeName='KmsMasterKeyId',
+                    AttributeValue=key
+                )
+            except ClientError as e:
```

Comment:
blanket exceptions are sketchy, ideally this is catching specific exceptions and let the rest bubble up.

Reply:
originally i wanted to catch invalid inputs, e.g. setting an invalid kms key as the key but it looks like it doesn't do any validation on the key itself.